### PR TITLE
Copy native runtime code to OSS.

### PR DESCRIPTION
**Pull request by zhxchen17**
Summary:

# High level context

Torch native runtime (codename: Sigmoid) is a new feature we're upstreaming to libtorch. Native runtime will execute a graph based on torch.export() format directly and the entire runtime is written in C++ so there's no Python dependency for running this.

The open sourcing part of the work is a follow up to our previous discussion in the doc https://docs.google.com/document/d/1KzZvJAscqKZdNaAmM3Is1A-VVgXSUclu-3Luk8z_m0k/edit?tab=t.0#heading=h.raww8l713hqp.

Since we have a very small C++ API surface (mostly on ModelRunner) and codebase, we prefer to copy the entire directory over in one diff to minimize the complications from syncing multiple diffs and redundant CI tests.

# Code layout

There are mainly two parts of sigmoid: C++ part and Python part. 
- Currently we put all the C++ code under `torch/csrc/nativert`. All the function/classes are put under c++ namespace `torch::nativert`
- Python code are put under `torch/export/experimental/`. Python code is needed for packaging torch.export() artifacts into *PT2 Archive* and later loaded by the C++ runtime.

# Supported Platform

Native Runtime is designed to work across the platforms meaning we support Linux, Windows and Mac by default. 
As one advanced feature, AOTI delegate will only work on Linux due to the limitation from Inductor.

# Core API

Frontend
```Python
ep = torch.export.export(model, args, kwargs)

from torch.export.experimental.package import package_model
from torch.export.experimental.package.pt2_archive import PT2ArchiveWriter

with open(FILENAME) as f:
    with PT2ArchiveWriter(f) as archive_writer:
        package_model(
            ep,
            MODEL_NAME,
            archive_writer,
        )
```

Backend
```c++
#include <torch/csrc/nativert/ModelRunner.h>

using torch::nativert::BaseRuntimeConfig;
using torch::nativert::ExecutorType;
using torch::nativert::Placement;
using torch::nativert::ModelRunner;

int main() {
  ExecutorType executor_type = ExecutorType::INTERPRETER;
  BaseRuntimeConfig runtime_config; // can be set to default value most of the time.
  Placement placement;  // can be set to default value most of the time.
  ModelRunner model_runner(FILENAME, MODEL_NAME, executor_type, runtime_config, placement);

  // args and kwargs should have the same PyTree spec as the example inputs from Python
  std::vector<c10::IValue> args = ...;
  std::unordered_map<std::string, c10::IValue> kwargs = ...;

  c10::IValue result = model_runner.run(args, kwargs); // safe to be called from multiple threads

  return 0;
}
```

Therefore here is the list of types we expose as TORCH_API:
- ModelRunner (and its base class ModelRunnerBase) (from torch/csrc/nativert/ModelRunner.h)
  - The main execution engine and entry point for exported graph.
- BaseRuntimeConfig (from torch/csrc/nativert/executor/ModelRunnerBase.h)
  - Centralized type to store configs/knobs.
- ExecutorType (from torch/csrc/nativert/executor/ModelRunnerBase.h)
  - The delegate executor type to be used for execution. Currently supports: 1. Plain interpreter (default) 2. AOTInductor 3. MTIA delagate.
- Placement (from torch/csrc/nativert/executor/Placement.h)
  - Typically not used. Sometimes useful for adjusting the device of weights during model loading.

# Test Plan

We already have a comprehensive test bundled with this diff.

The basic idea is that we automatically generate tests based on current unittests of torch export (https://fburl.com/pikxf3xi). There are 340 different model code from torch export's test suite and sigmoid is passing on 98% of them.

To run the unittests of sigmoid:
```
pytest test/export/test_nativert.py
```

In addition to the unittests we bundle in OSS, we also have many tests running inside fbcode, but we will stick with export's unittest for a while in OSS.

# Shared utils from c10/aten

I believe in fact we have very little duplicated code with libtorch's current codebase. The reason is that sigmoid is developed to be tightly integrated with libtorch from the beginning and we took opportunities to reuse everything we could from libtorch from day 1. Examples include:
- c10::IValue
- c10::Synchronized
- C10_LIKELY and TORCH_CHECK macros
- c10::FastMap
- ATen/core/function_schema
- ATen/core/dispatch/Dispatcher

For the sake of development speed we made the best effort to reuse libtorch as much as we could and maintaining the graph execution engine on top of this.

Certainly there are some exceptions. If something is not stable enough in the core to be reused, we would implement it on our own. The best example is `torch/csrc/nativert/executor/AOTInductorModelImpl.cpp` which could be refactored to merge with at::AOTIModelContainerRunner in the future (since this part is also in active development in core and compiler team have control over both of them).

### CODE REVIEW

File: build_variables.bzl [hunk 1]
```diff
@@ -619,7 +656,7 @@ libtorch_lite_cmake_sources = sorted(
     torch_mobile_core,
 )
 
-libtorch_cmake_sources = libtorch_core_sources + libtorch_core_jit_sources
+libtorch_cmake_sources = libtorch_core_sources + libtorch_core_jit_sources + libtorch_nativert_sources
```
Comment (gmagogsfm) [inline build_variables.bzl:659]: Did you measure binary size increase for libtorch? What does it look like?

Comment (gmagogsfm) [reply build_variables.bzl:659]: Looks good, quite a small increase

**Review by gmagogsfm — APPROVED** [review]

**Comment by zhxchen17** [comment]
Since this diff is a bit large to review, @albanD suggested to put more context on on the following aspects:
- New APIs, their behavior and documentation
- Testing to make sure they are behaving properly and everything is covered
- Shared utils to make sure we dedup everything that can be dedupped
- Misc, make sure lint, codemod, build, etc are all done properly.

I will work on making a better PR description next.

**Comment by seemethere** [comment]
Curious how much test time this adds to the existing test suite.

File: torch/export/experimental/package/pt2_archive.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        data = self.read_bytes(name)
+        return data.decode()
+
+    def archive_version(self) -> int:
+        """
+        Get the archive version.
+        """
+        try:
+            archive_version = self.read_string(ARCHIVE_VERSION_PATH)
+        except Exception:
```
Comment (albanD) [inline torch/export/experimental/package/pt2_archive.py:177]: I assume lint was complaining about this. We need a tighter exception handling.

**Comment by albanD** [comment]
I was only able to look at a couple files because github UI keeps crashing, this is by no means a full review.

File: .lintrunner.toml [hunk 3]
```diff
@@ -93,6 +93,8 @@ exclude_patterns = [
     'torch/csrc/inductor/aoti_torch/generated/**',
     'torch/csrc/jit/serialization/mobile_bytecode_generated.h',
     'torch/csrc/utils/pythoncapi_compat.h',
+    'torch/csrc/nativert/common/lightweightsemaphore.h',
```
[the remark this answers is not in the document]
Comment (zhxchen17) [reply .lintrunner.toml:96]: These two files are copied from a third party library https://github.com/cameron314/concurrentqueue, which has zlib license and we think it might be better to just exclude them from the linters since we didn't author the code.

File: test/export/test_nativert.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    PyRuntimeConfigs,
+)
+from torch._subclasses.fake_tensor import FakeTensor
+from torch.export.experimental.package import package_model
+from torch.export.experimental.package.pt2_archive import PT2ArchiveWriter
+from torch.utils import _pytree as pytree
+
+
+try:
+    from . import test_export, testing
```
[the remark this answers is not in the document]
Comment (zhxchen17) [reply test/export/test_nativert.py:23]: Looks like we were doing this all over the place https://github.com/pytorch/pytorch/blob/main/test/export/test_retraceability.py#L4

File: test/export/test_nativert.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+]
+for test in tests:
+    make_dynamic_cls(test, strict=True)
+    make_dynamic_cls(test, strict=False)
+del test
+
+if __name__ == "__main__":
+    from torch._dynamo.test_case import run_tests
+
+    run_tests()
```
[the remark this answers is not in the document]
Comment (zhxchen17) [reply test/export/test_nativert.py:220]: If you were asking how do I check there are tests being run, I just use `pytest test/export/test_nativert.py`

**Review by malfet — COMMENTED** [review]
`torch/csrc` feels like a weird place to put native runtime into, as name of the folder implies that code there is CPython bindings. What's wrong with `nativert` top level folder?

File: test/export/test_nativert.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/export/test_nativert.py:23]: We should fix that, because that would mean we run the same test many many times :/

File: test/export/test_nativert.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/export/test_nativert.py:220]: Given that this is non-standard test setup, we should ensure that the test that do run in CI are the ones you expect.

File: .lintrunner.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (malfet) [reply .lintrunner.toml:96]: All code vendored as-is from 3rd-party library should be located in `third_party` folder (for C++/CUDA/Rust) and `torch/_vendor/` for Python (see miniz integration as an example)
